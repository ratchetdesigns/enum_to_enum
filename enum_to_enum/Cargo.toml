[package]
name = "enum_to_enum"
version = "0.1.0"
authors = ["Adam Berger <adam@ratchetdesigns.com>"]
license = "MIT"
description = "Derives possibly effectful conversions between enums"
include = ["README.md", "LICENSE-MIT", "src/**/*.rs"]
edition = "2018"

[dependencies]
enum_to_enum_derive = { version = "=0.1.0", path = "../enum_to_enum_derive"}
