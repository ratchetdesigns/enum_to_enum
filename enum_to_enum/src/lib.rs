#![warn(missing_docs)]

//! # enum_to_enum
//!
//! enum_to_enum exposes a derive macro to easily generate possibly effectful enum-to-enum conversions.
//!
//! The destination enum must be annotated to specify one or more source enums we will generate
//! [`From`] implementations for.
//!
//! The generated `From` implementations rely on provided [`From`] or [`TryFrom`](std::convert::TryFrom) implementations for
//! each corresponding field or tuple-item in struct-like or tuple-like enums.
//!
//! Each variant of the destination enum may specify one or more variants of the source enums that
//! should correspond to the destination variant. If multiple variants of a given source enum might
//! correspond to the same destination variant, the destination variant must have at least 1 field
//! or tuple item and the corresponding `TryFrom` implementations will be invoked in the order in
//! which they appear on the destination enum until one of them succeeds.
//!
//! Effectful conversions require users to provide a struct implementing the [`WithEffects`] trait
//! and a conversion will be generated from each source enum to the provided `effect_container`.

#[allow(unused_imports)]
#[macro_use]
extern crate enum_to_enum_derive;

pub use enum_to_enum_derive::*;

/// Any struct specified as an `effect_container` for the [`from_enum`](enum_to_enum_derive::FromEnum) attribute must implement `WithEffects`.
/// `WithEffects` specifies a container for a value, the result of some conversion, and an ordered
/// list of effects arising from that conversion.
///
/// It is expected that these effects correspond to some actions to take subsequent to the
/// conversion.
pub trait WithEffects
where
    Self: Sized,
{
    /// The type of the value we store. This is the result of the conversion.
    type Value;

    /// The type of the effects we store.
    type Effect;

    /// Creates a new instance of this `WithEffects` implementor with the given value and effects.
    fn new(value: Self::Value, effects: Vec<Self::Effect>) -> Self;

    /// Converts self into a tuple of its value and an iterator over its effects.
    fn into_value_and_effects(self) -> (Self::Value, Box<dyn Iterator<Item = Self::Effect>>);

    /// Creates a new instance of this `WithEffects` implementor with the provided value and
    /// effects.
    /// We provide a default implementation in terms of [`new`](Self::new).
    fn compose_from(value: Self::Value, composed_effects: Box<[Self::Effect]>) -> Self {
        let effects = composed_effects.into();
        Self::new(value, effects)
    }
}
