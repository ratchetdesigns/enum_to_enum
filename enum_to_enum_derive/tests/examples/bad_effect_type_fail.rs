use enum_to_enum::FromEnum;

enum Src {
    Case1(String),
}

#[derive(FromEnum)]
#[from_enum(Src, effect_container = BadEffectHolder)]
enum EffectDest {
    Case1(String),
}

struct BadEffectHolder<V> {
    v: V,
}

fn main() {}
