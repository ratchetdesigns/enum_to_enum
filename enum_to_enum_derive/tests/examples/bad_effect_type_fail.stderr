error[E0599]: no method named `into_value_and_effects` found for struct `BadEffectHolder` in the current scope
  --> $DIR/bad_effect_type_fail.rs:8:37
   |
8  | #[from_enum(Src, effect_container = BadEffectHolder)]
   |                                     ^^^^^^^^^^^^^^^ method not found in `BadEffectHolder<String>`
...
13 | struct BadEffectHolder<V> {
   | ------------------------- method `into_value_and_effects` not found for this
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `into_value_and_effects`, perhaps you need to implement it:
           candidate #1: `WithEffects`

error[E0277]: the trait bound `BadEffectHolder<String>: From<String>` is not satisfied
 --> $DIR/bad_effect_type_fail.rs:7:10
  |
7 | #[derive(FromEnum)]
  |          ^^^^^^^^ the trait `From<String>` is not implemented for `BadEffectHolder<String>`
  |
  = help: the following implementations were found:
            <BadEffectHolder<EffectDest> as From<Src>>
  = note: required because of the requirements on the impl of `Into<BadEffectHolder<String>>` for `String`
  = note: this error originates in the derive macro `FromEnum` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: no function or associated item named `compose_from` found for struct `BadEffectHolder` in the current scope
  --> $DIR/bad_effect_type_fail.rs:8:37
   |
8  | #[from_enum(Src, effect_container = BadEffectHolder)]
   |                                     ^^^^^^^^^^^^^^^ function or associated item not found in `BadEffectHolder<_>`
...
13 | struct BadEffectHolder<V> {
   | ------------------------- function or associated item `compose_from` not found for this
   |
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the following trait defines an item `compose_from`, perhaps you need to implement it:
           candidate #1: `WithEffects`

error[E0277]: the trait bound `BadEffectHolder<V>: WithEffects` is not satisfied
 --> $DIR/bad_effect_type_fail.rs:8:37
  |
8 | #[from_enum(Src, effect_container = BadEffectHolder)]
  |                                     ^^^^^^^^^^^^^^^
  |                                     |
  |                                     the trait `WithEffects` is not implemented for `BadEffectHolder<V>`
  |                                     required by this bound in `assert_implements_with_effects`
